service/src/main/java/org/mskcc/smile/service/impl/TempoMessageHandlingServiceImpl.java

```diff
@@ -364,63 +367,79 @@ public void run() {
         }
     }
 
-    private void publishTempoSamplesToCBioPortal(Set<String> tumorPrimaryIds) throws Exception {
+    private class TempoEmbargoStatusHandler implements Runnable {
+        final Phaser phaser;
+        boolean interrupted = false;
+
+        TempoEmbargoStatusHandler(Phaser phaser) {
+            this.phaser = phaser;
+        }
+
+        @Override
+        public void run() {
+            phaser.arrive();
+            while (true) {
+                try {
+                    List<String> samplePrimaryIds = tempoEmbargoStatusQueue.poll(100, TimeUnit.MILLISECONDS);
+                    if (samplePrimaryIds != null) {
+                        LOG.info("Updating access level for " + samplePrimaryIds.size() + " samples...");
+                        tempoService.updateTempoAccessLevel(samplePrimaryIds,
+                                TempoServiceImpl.ACCESS_LEVEL_PUBLIC);
+                        // publish to tempo sample update topic
+                        publishTempoSamplesToCBioPortal(new HashSet<>(samplePrimaryIds),
+                                TEMPO_UPDATE_SAMPLES_EMBARGO_TOPIC);
+                    }
+                } catch (InterruptedException e) {
+                    interrupted = true;
+                } catch (Exception e) {
+                    LOG.error("Error during handling of sample billing data", e);
+                }
+                tempoEmbargoStatusHandlerShutdownLatch.countDown();
+            }
+        }
+    }
+
+    private void publishTempoSamplesToCBioPortal(Set<String> samplePrimaryIds, String topic)
+            throws Exception {
         // validate and build tempo samples to publish to cBioPortal
         Set<TempoSample> validTempoSamples = new HashSet<>();
-        for (String primaryId : tumorPrimaryIds) {
+        LOG.info("Assembling TEMPO data per sample for publishing....");
+        for (String primaryId : samplePrimaryIds) {
             try {
+                TempoSample tempoSample = tempoService.getTempoSampleDataBySamplePrimaryId(primaryId);
```

Comment:
I like how this query queries all the Tempo sample data in one call. To take it further, what do you think about querying **all** the Tempo samples in one call to Neo4j by passing in all the sample Primary IDs?

Reply:
discussed